Refactor lookup functions

lpw25 opened the pull request:
This PR refactors the lookup functions used for searching the environment for a `Longident.t`:

- It moves all the code from the `Typetexp.find_*` functions into the coressponding `Env.lookup_*` functions. This puts all the handling of lookup errors, deprecation warnings and usage warnings into a single place.
- The `lookup_*` functions now always take a location
- The `lookup_*` functions now take an optional `use` parameter indicating whether this should be considered a use of the item from the perspective of usage or deprecation warnings.
- For cases that should not cause an error, I've added `Env.find_*_by_name` functions that raise `Not_found` rather than an error if the item can't be found.
- Other minor cleanups around looking things up from the environemnt...

These changes are not a complete no-op. They change the locations for constructor usage warnings, improve the semantics of constructor usage warnings for private constructors, and they improve a couple of error messages related to instance variables. I've added some tests to make sure that all these changes are demonstrated.

lpw25 committed:
Refactor environment lookup functions

lpw25 committed:
Add more unused constructor tests

lpw25 committed:
Add some tests related to instance variables

alainfrisch commented:
The suggested changes look great, and after a first quick review pass, I don't have much comments on the commits.  Two remarks:

  - What about returning options instead of raising Not_found for *_by_name functions?

  - To help understand the impact on error messages, it would be good to add the new expect tests in a commit before the actual changes (of course, the reviewers could just copy/paste the tests in a toplevel...).

lpw25 commented:
> What about returning options instead of raising Not_found for *_by_name functions?

I'd prefer to keep the `find_*_by_name` functions consistent with the `find_*` functions. So either both should return options or both should use `Not_found`.

> To help understand the impact on error messages, it would be good to add the new expect tests in a commit before the actual changes (of course, the reviewers could just copy/paste the tests in a toplevel...).

I did.

lpw25 commented:
> I did

Possibly Github's interface doesn't make this clear because the test commits have a later commit date than the main commit, but they are in fact before the main commit in the history.

alainfrisch commented on testsuite/tests/typing-warnings/unused_types.ml:
```diff
@@ -117,11 +117,6 @@ end = struct
 end
 ;;
 [%%expect {|
-Line 5, characters 2-20:
```
Isn't this one a regression? It seems the warning was right that the constructor is never used to build values.

lpw25 replied:
This is a deliberate change. The warning doesn't make sense for a constructor that was defined as `private`. Obviously values of that type are never constructed: it was defined as an empty type. 

Occasionally people use such types to make empty types for GADT indices that are known to be incompatible. Currently if they do this locally to a module they get a warning that cannot be silenced without using attributes, this allows the warning to be reasonably silenced.

Note that this was already the case if the type was exported as a private type. So code like:
```ocaml
module M = sig type t = private T end = struct type t = private T end
```
specifically doesn't raise the warning about the declaration in the `struct`. However this case is completely isomorphic to the other one -- we still know that no-one is going to create a value with this constructor.

alainfrisch replied:
Ok, it makes sense, but the rationale is not straightforward and writing it down (perhaps just as a comment somewhere in the code) would be useful for future reference.  (Sorry if this is already documented.)

alainfrisch commented on testsuite/tests/typing-warnings/unused_types.ml:
```diff
@@ -285,11 +280,6 @@ end = struct
 end
 ;;
 [%%expect {|
-Line 6, characters 20-31:
```
Same here.

alainfrisch commented:
Right, sorry.  Indeed I was confused by the UI.

alainfrisch commented:
check-typo failures:

./testsuite/tests/typing-objects/Tests.ml:950.1: [white-at-eof] empty line(s) at EOF
./typing/env.ml:78.81: [long-line] line is over 80 columns
./typing/typecore.ml:750.81: [long-line] line is over 80 columns
./typing/typecore.ml:901.81: [long-line] line is over 80 columns
./typing/typecore.ml:2833.81: [long-line] line is over 80 columns
./typing/typecore.ml:2936.81: [long-line] line is over 80 columns

lpw25 commented:
Fixed the check-typo failures.

trefis commented on typing/untypeast.ml:
```diff
@@ -107,11 +107,8 @@ let map_loc sub {loc; txt} = {loc = sub.location sub loc; txt}
-    with
-      | Not_found -> aux (i+1)
+    if Env.bound_value name env then aux (i+1)
+    else name
```
The behavior is clearly changing here, but this looks like a bugfix.

alainfrisch replied:
What about adding a test that illustrates the fix?

trefis commented on typing/typetexp.ml:
```diff
@@ -28,8 +28,7 @@ exception Already_bound
     Unbound_type_variable of string
-  | Unbound_type_constructor of Longident.t
-  | Unbound_type_constructor_2 of Path.t
+  | Undefined_type_constructor of Path.t
```
:heart:

trefis commented on typing/env.ml:
```diff
@@ -224,27 +220,30 @@ module TycompTbl =
+      else begin
+        match using with
+        | None -> nothing
+        | Some f ->
```
I'm not a fan of this new version. I think I'd rather change `find_all` to pass `None` when `mark = false`

trefis commented on typing/env.ml:
```diff
@@ -224,27 +220,30 @@ module TycompTbl =
+      end
 
-    let rec find_all name tbl =
+    let rec find_all ~mark name tbl =
```
Btw, I don't think this function is ever called with `~mark:false`.
The API might be better / more consistent this way, but I'm not sure I'm convinced.

trefis commented on bytecomp/matching.ml:
```diff
@@ -1480,12 +1480,13 @@ let get_mod_field modname field =
+        | _ ->
+            fatal_error ("Primitive "^modname^"."^field^" not found.")
+        | exception Not_found ->
+            fatal_error ("Primitive "^modname^"."^field^" not found.")
```
Good news: these two cases can now be merged!

(Well, except that once you rebase the code doesn't look like that anymore, and that the function should probably disappear anyway)

trefis commented on typing/typetexp.ml:
```diff
@@ -74,149 +55,6 @@ module TyVarMap = Misc.Stdlib.String.Map
-
-(* Narrowing unbound identifier errors. *)
-
-let rec narrow_unbound_lid_error : 'a. _ -> _ -> _ -> _ -> 'a =
```
Very glad to see this go. The new way you propagate errors is much better than the replaying of various operations that was going on here.

trefis commented on typing/env.ml:
```diff
@@ -885,7 +934,8 @@ let rec find_module_descr path env =
       begin match get_components (find_module_descr p1 env) with
         Functor_comps f ->
-          !components_of_functor_appl' f env p1 p2
+          let loc = Location.(in_file !input_name) in
```
I was about to question that `Location.in_file`, asking if there isn't a better location we could propagate down to here. As it turns out, it has simply been pulled out of `components_of_functor_appl`, and the other calls do pass more precise locations. So I guess it's fine.

trefis commented on typing/ctype.ml:
```diff
@@ -3866,18 +3864,8 @@ let rec filter_visited = function
-  | Path.Pdot (p1, s, _) ->
-      Longident.Ldot (lid_of_path p1, hash ^ s)
-  | Path.Papply (p1, p2) ->
-      Longident.Lapply (lid_of_path ~hash p1, lid_of_path p2)
```
Passing `~hash` when recursing on `p1` was just nonsense, wasn't it?

trefis commented on typing/env.mli:
```diff
@@ -90,52 +95,130 @@ val add_required_global: Ident.t -> unit
+val bound_type: string -> t -> bool
+val bound_modtype: string -> t -> bool
+val bound_class: string -> t -> bool
+val bound_cltype: string -> t -> bool
```
FTR: it looks like only `bound_value` is ever used in the codebase. But I'd keep the functions here: giving compiler-libs a decent API seems good.

trefis reviewed (commented):
As the code began to rot, I also gave this a first round of review.
I left a few notes here and there, but overall I agree with Alain: this looks good.

If you manage to rebase in the next few days, I'll be happy to finish reviewing and approve this.

PS: check_typo is still unhappy.

lpw25 commented:
I've rebased this and addressed review comments.

Some parts of the rebase pushed parts of the code past my personal complexity threshold. So I decided to make some other changes to `Env`'s internals to try to make the code simpler. I replaced some tuples with records, and made some parts of the environment use the same types instead of gratuitously different ones. I think I've managed to sufficiently decrease the amount of duplicated complex code through these changes.

trefis commented on Makefile.dev:
```diff
@@ -45,4 +45,4 @@ list-all-asts:
 
 partialclean::
-	rm -f $(AST_FILES)
+	@rm -f $(AST_FILES)
```
This shouldn't be here.

lpw25 replied:
removed

trefis commented on typing/env.mli:
```diff
@@ -109,47 +126,125 @@ val add_required_global: Ident.t -> unit
+  ?use:bool -> loc:Location.t -> string -> t ->
+  Path.t * Asttypes.mutable_flag * string * type_expr
+
+(* Lookup by long identifiers. Raise [Not_found] if there would be an error. *)
```
It's not clear from the mli how the `find_XXX_by_name` functions are different from `lookup_XXX`.
This is linked to the fact that we don't know what `lookup_XXX` raises, but also that don't know whether `find_...` is going to count as an use or not.

lpw25 replied:
I've added a comment to make things clearer

trefis commented on typing/env.ml:
```diff
@@ -677,180 +798,194 @@ let reset_cache_toplevel () =
   | _ ->
       assert false
 
 let find_type_full path env =
```
I know this was already the case before, but since you're rewriting that whole area: couldn't this function be given a different name?

lpw25 replied:
I couldn't think of a good name so I've left it like this for now

trefis commented on typing/typemod.ml:
```diff
@@ -1479,18 +1481,11 @@ and transl_recmodule_modtypes env sdecls =
-         Env.add_module ~arg:true id Mp_present dummy env
+         (* cf #5965 *)
+         Env.enter_unbound_module (Ident.name id)
+           Mod_unbound_illegal_recursion env
```
:heart:

trefis commented on typing/types.ml:
```diff
@@ -104,11 +104,6 @@ and value_kind =
-
-and value_unbound_reason =
-  | Val_unbound_instance_variable
-  | Val_unbound_ghost_recursive
```
:heart:

trefis commented on typing/env.mli:
```diff
@@ -109,47 +126,125 @@ val add_required_global: Ident.t -> unit
+  | No_hint
+  | Missing_rec of Location.t
+
+type lookup_error =
```
I believe the `lookup_*` functions above raise `Error (Lookup_error _)`  with `_` being a value of this type.
It'd be nice if that was indicated somewhere in this mli.

trefis commented on typing/env.ml:
```diff
@@ -584,54 +684,75 @@ let find_same_module id tbl =
+let find_name_module ~mark name tbl =
+  match IdTbl.find_name wrap_module ~mark name tbl with
+  | x -> x
+  | exception Not_found when not (Current_unit_name.is name) ->
```
Shouldn't this be `when not (Current_unit_name.is name) && !Clflags.transparent_modules`?

lpw25 replied:
I don't think so. That was a weird discrepancy between `lookup_module` and `lookup_module_descr` in the old code. It meant that in theory `M.t` could sometimes be resolved when `M` could not, although in practice I don't think this would have happened except in very strange corner cases involving parallel builds. The new code is more obviously consistent.

trefis reviewed (commented):
I looked at this once more, apart from the minor comments I left here and there, there are two things that bother me:
- I believe there might be semantic changes that are not described anywhere.
This refers to the uses of `find_module_by_name`, `lookup_module` and `lookup_module_path` (and the various values of the `load` parameter), which I don't think line up with the previous version of the code.
Can you comment?
- I'm not sure when to use the new `find_xxx_by_name` functions, compared to `lookup_xxx`: there's nothing in the API/types helping me make that choice, and nothing in the doc either.
In this PR you say "use them when there shouldn't be an error", and I could almost buy it: "OK, let's just use this in printtyp, and when I know for sure the type is there" (e.g. `selfpat-*`).
But if there won't be an error, why should I care anyway?
Also, I can see that `approx_type` uses `find_type_by_name`, but that `approx_modtype` uses `lookup_modtype ~use:false`; I know there was already a similar difference in the previous version of the code, I'm just pointing out that the principle you stated might not be the right one.

Drup commented:
I didn't do a proper review, but I had a quick look at the new API, and I like it quite a lot. It's a clear improvement over the previous one.

lpw25 committed:
Bootstrap

lpw25 commented:
I've rebased again and updated based on the review comments.

> I believe there might be semantic changes that are not described anywhere.
> This refers to the uses of find_module_by_name, lookup_module and lookup_module_path (and the various values of the load parameter), which I don't think line up with the previous version of the code.
> Can you comment?

Mostly, `Typetexp.find_foo` has been replaced by `Env.lookup_foo` and direct uses of `Env.lookup_foo` have been replaced by `Env.find_foo_by_name` (or `Env.bound_foo` if the result was ignored). The exceptions are:

- The `Typetexp.find_foo`s in `Typemod.approx_modtype` have been replaced by `lookup_foo ~use:false`. In principle, this changes the warnings behaviour, although the existing use of `Warnings.without_warnings` probably makes it a no-op. The change is correct because such uses are going to be looked up again when we typecheck the module types properly later.

- The `Env.lookup_type` in `Typecore.approx_type` is replaced by `Env.lookup_type ~use:false`. This changes the error behaviour of this lookup. I believe that it is correct since the right behaviour when the type is not bound is to emit an error.

- The `Env.lookup_value` in `Typecore.type_expression` is replaced by `Env.lookup ~use:false`. In principle, this changes the error behaviour, although in practice we never get this far in the case of an error. It seemed more correct to use `lookup_value` than `find_value_by_name` because if somehow we couldn't find the value then the right behaviour would be to emit a proper error.

In addition, the uses of `lookup_module_descr ~mark:true` in `Env.find_all` and `Env.find_all_simple_list` are replaced by `lookup_module_components ~errors:false ~use:false`. This changes the marking behaviour of these functions. The change is correct since these functions are only used for spellchecking hints and should not be marking things as used.

trefis reviewed (approved):
I believe I have reviewed everything, and it all makes sense to me.

Please add a changelog entry and then this should be good to merge.

trefis commented:
This is going to be a pain to rebase, and I don't want to review it again in 6 months. Apparently I can't  push to @lpw25's repository, so I'll just merge this now and add the changelog entry after the fact.

trefis commented:
Changelog entry added in f69e08e.

hhugo commented on typing/typetexp.ml:
```diff
@@ -415,8 +253,7 @@ and transl_type_aux env policy styp =
   | Ptyp_class(lid, stl) ->
       let (path, decl, _is_variant) =
         try
-          let path = Env.lookup_type lid.txt env in
```
I believe this change introduced the #9170

gasche commented:
Note from the future: this PR (merged in 4.10) removed the `Env.fold_*` functions (by moving their user, the "Did you mean ..." spellchecking hint, inside the Env module), but it turns out that language tooling needs them: they were added again by Merlin maintainers in #9922, and I just learned that [ocamlearlybird](https://github.com/hackwaly/ocamlearlybird/) also relied on them.
